### PR TITLE
[Serve] Pin the fastapi & starlette version to avoid breaking proxy

### PR DESCRIPTION
**Pull request by sihanwang41**
## Why are these changes needed?

1. Pin the fastapi version to 0.108.0. The latest version of fastapi is breaking with starlette. (issue) 
2. Add starlette compact for supporting >= 0.35.0 version.

Test Manually:

By running (fastapi == 0.108.0): 
```
import ray
import requests
from fastapi import FastAPI
from ray import serve

app = FastAPI()


@serve.deployment
@serve.ingress(app)
class MyFastAPIDeployment:
    @app.get("/")
    def root(self):
        return "Hello, world!"


serve.run(MyFastAPIDeployment.bind(), route_prefix="/hello")
resp = requests.get("http://localhost:8000/hello")
# Got 404 response
assert resp.json() == "Hello, world!"
```

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by edoakes** [comment]
@sihanwang41 not following the description -- why can't we fix the issue instead of pinning?

**Comment by sihanwang41** [comment]
> @sihanwang41 not following the description -- why can't we fix the issue instead of pinning?

Hi @edoakes, right now, there is another issue (paste link in the pr) with latest fastapi version, it is breaking the serve fastapi test. (Our ci is using pinned version, which is not able to catch it).

pip install fastapi==0.109.0
```
import ray
import requests
from fastapi import FastAPI
from ray import serve

app = FastAPI()


@serve.deployment
@serve.ingress(app)
class MyFastAPIDeployment:
    @app.get("/")
    def root(self):
        return "Hello, world!"


serve.run(MyFastAPIDeployment.bind(), route_prefix="/hello")
resp = requests.get("http://localhost:8000/hello")
assert resp.json() == "Hello, world!"
```

The previous version of fastapi works fine. So that is why I pin the fastapi version. I probably don't need to pin starlette version, since the previous fastapi version is using the "good" starlette version, so there is no `AttributeError: 'Middleware' object has no attribute 'options'` error.

**Comment by sihanwang41** [comment]
@edoakes PTAL, thx!

File: python/ray/serve/_private/proxy.py [hunk 1]
```diff
@@ -1211,9 +1211,17 @@ def __init__(
         self.wrapped_http_proxy = self.http_proxy
 
         for middleware in http_middlewares:
-            self.wrapped_http_proxy = middleware.cls(
-                self.wrapped_http_proxy, **middleware.options
-            )
+            if hasattr(middleware, "options"):
```
[the remark this answers is not in the document]
Comment (sihanwang41) [reply python/ray/serve/_private/proxy.py:1214]: good point

**Review by GeneDer — APPROVED** [review]
LGTM, thanks Sihan!

File: python/ray/serve/_private/proxy.py [hunk 2]
```diff
@@ -1211,9 +1212,17 @@ def __init__(
         self.wrapped_http_proxy = self.http_proxy
 
         for middleware in http_middlewares:
-            self.wrapped_http_proxy = middleware.cls(
-                self.wrapped_http_proxy, **middleware.options
-            )
+            if starlette.__version__ <= "0.34.0":
```
Comment (edoakes) [inline python/ray/serve/_private/proxy.py:1215]: This should use `packaging.version` for comparison

**Review by edoakes — COMMENTED** [review]
Please update the PR description to have the full context.

What is the testing plan?

File: python/ray/serve/_private/proxy.py (hunk 2 above)
Comment (sihanwang41) [reply python/ray/serve/_private/proxy.py]: done.

**Comment by sihanwang41** [comment]
> Please update the PR description to have the full context.
> 
> What is the testing plan?

manually tested with 0.108.0 fastapi version. and Update the test-requirements.txt

**Review by edoakes — APPROVED** [review]
LGTM

Can you please create a tracking issue for us to unpin the fastapi version as soon as they've fixed the problem upstream?

**Comment by sihanwang41** [comment]
merge the latest master doesn't help, still having irrelevant failure.

merge again.

**Comment by edoakes** [comment]
Current failures are caused by a merge conflict on @zcin's PRs. Will be addressed shortly: https://github.com/ray-project/ray/pull/42718

**Pull request merged**